zebra-chain/src/parameters/network.rs

```diff
@@ -51,52 +53,112 @@ mod tests;
+    Regtest,
 }
 
-use zcash_primitives::consensus::{Network as ZcashPrimitivesNetwork, Parameters as _};
-impl Network {
+impl From<Network> for NetworkKind {
+    fn from(network: Network) -> Self {
+        network.kind()
+    }
+}
+
+/// An enum describing the possible network choices.
+#[derive(Clone, Debug, Default, Eq, PartialEq, Hash, Serialize)]
+#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+#[serde(into = "NetworkKind")]
+pub enum Network {
+    /// The production mainnet.
+    #[default]
+    Mainnet,
+
+    /// A test network such as the default public testnet,
+    /// a configured testnet, or Regtest.
+    Testnet(Arc<NetworkParameters>),
+}
+
+impl NetworkKind {
     /// Returns the human-readable prefix for Base58Check-encoded transparent
     /// pay-to-public-key-hash payment addresses for the network.
-    pub fn b58_pubkey_address_prefix(&self) -> [u8; 2] {
-        <ZcashPrimitivesNetwork>::from(self).b58_pubkey_address_prefix()
+    pub fn b58_pubkey_address_prefix(self) -> [u8; 2] {
+        match self {
+            Self::Mainnet => constants::mainnet::B58_PUBKEY_ADDRESS_PREFIX,
+            Self::Testnet | Self::Regtest => constants::testnet::B58_PUBKEY_ADDRESS_PREFIX,
+        }
     }
 
     /// Returns the human-readable prefix for Base58Check-encoded transparent pay-to-script-hash
     /// payment addresses for the network.
-    pub fn b58_script_address_prefix(&self) -> [u8; 2] {
-        <ZcashPrimitivesNetwork>::from(self).b58_script_address_prefix()
-    }
-    /// Returns true if the maximum block time rule is active for `network` and `height`.
-    ///
-    /// Always returns true if `network` is the Mainnet.
-    /// If `network` is the Testnet, the `height` should be at least
-    /// TESTNET_MAX_TIME_START_HEIGHT to return true.
-    /// Returns false otherwise.
-    ///
-    /// Part of the consensus rules at <https://zips.z.cash/protocol/protocol.pdf#blockheader>
-    pub fn is_max_block_time_enforced(&self, height: block::Height) -> bool {
+    pub fn b58_script_address_prefix(self) -> [u8; 2] {
         match self {
-            Network::Mainnet => true,
-            Network::Testnet => height >= super::TESTNET_MAX_TIME_START_HEIGHT,
+            Self::Mainnet => constants::mainnet::B58_SCRIPT_ADDRESS_PREFIX,
+            Self::Testnet | Self::Regtest => constants::testnet::B58_SCRIPT_ADDRESS_PREFIX,
+        }
+    }
+
+    /// Return the network name as defined in
+    /// [BIP70](https://github.com/bitcoin/bips/blob/master/bip-0070.mediawiki#paymentdetailspaymentrequest)
+    pub fn bip70_network_name(&self) -> String {
+        if *self == Self::Mainnet {
+            "main".to_string()
+        } else {
+            "test".to_string()
+        }
+    }
+}
+
+impl From<NetworkKind> for &'static str {
+    fn from(network: NetworkKind) -> &'static str {
+        // These should be different from the `Display` impl for `Network`
```

Comment:
> Is it so that we don't confuse NetworkKind with Network in logs?

Yeah, I was thinking so its lowercase form can't be parsed as the default Testnet in the `FromStr` trait impl and so it's generally harder to confuse the two, distinguishing them in logs is a good reason too, I'll add both.

Do you think we could also remove the `FromStr` trait impl, or do you know why we have it? I haven't seen it being used anywhere.

Reply:
> Do you think we could also remove the FromStr trait impl, or do you know why we have it?

We need it because we derive a command-line arg [`Parser`](https://docs.rs/clap/latest/clap/trait.Parser.html) here: https://github.com/zcashfoundation/zebra/blob/05e41a28037ca4288a61b7a7e9a4e0ce40e40b98/zebrad/src/commands/tip_height.rs#L21-L31, which needs to parse the network from a command-line arg. 

I guess the macro impls https://docs.rs/clap/latest/clap/builder/struct.ValueParser.html# for `Network` through https://docs.rs/clap/latest/clap/macro.value_parser.html#

I guess we could impl any other trait for `Network` that is listed among the "Supported types" here: https://docs.rs/clap/latest/clap/macro.value_parser.html#, but that probably wouldn't help anything.

We introduced the parser in #4289 as a Zebra utility, then moved it to a Zebra command.

Reply:
Added the reason to the comment and also a note about where the `FromStr` impl is used in https://github.com/ZcashFoundation/zebra/pull/8368/commits/86cd919135c40e5b2dd512c1ed14aa0963465262#diff-4dcd6cfa50ff2b87d1dc68673b1ec3a39b816a0473d15a17423a80ae80029bca.

Thank you for finding where that's used.